contents/java/Spring.md

```diff
@@ -0,0 +1,277 @@
+  * [Spring AOP](#spring-aop)
+* [Why use Spring‎](#why-use-spring)
+  * [Benefit 1: Modularity](#benefit-1-modularity)
+  * [Benefit 2: Easy Integration](#benefit-2-easy-integration)
+  * [Benefit 3: Strong Community Support](#benefit-3-strong-community-support)
+  * [Benefit 4: Usability](#benefit-4-usability)
+* [How to get started‎](#how-to-get-started)
+</box>
+
+## What is Spring Framework?
+
+The [official website](https://docs.spring.io/spring/docs/current/spring-framework-reference/overview.html) describes Spring Framework as follows:
+
+>*Spring Framework* is an open-source framework to create Java enterprise applications.
+>It provides foundational support for different application architectures, including messaging, transactional data and persistence, and web.
+
+## Core Features
+
+### The IoC Container
+
+Normally, objects in a Java application often have to take on additional responsibilities such as creating other objects,
+wiring them together, configuring them, and managing their life cycle. When using the Spring framework for an application,
+the framework takes care of many of those responsibilities. This design approach is called inversion of control (IoC)
+and Spring achieves it using something called the IoC container. The Spring IoC container manage the
+<tooltip content="A bean is an object that is instantiated, assembled, and otherwise managed by a Spring IoC container.">
+<i>beans </i>
+</tooltip>in Spring Framework.
+
+In the Spring framework, the IoC container is represented by the interface `ApplicationContext`.
+The beans are created with the configuration metadata that you supply to the container,
+which can be in the form of XML configuration or annotations.
+
+Below is one way to manually instantiate a container:
+```java
+ApplicationContext context = new ClassPathXmlApplicationContext("applicationContext.xml");
+```
+
+For example, say we have an Employee class like so:
+```java
+package com.company;  
+  
+public class Employee {  
+    private int id;  
+    private String name;  
+    private String city;  
+  
+    public int getId() {  
+        return id;  
+    }  
+    public void setId(int id) {  
+        this.id = id;  
+    }  
+    public String getName() {  
+        return name;  
+    }  
+    public void setName(String name) {  
+        this.name = name;  
+    }  
+    void display(){  
+        System.out.println(id+" "+name);  
+    }
+}        
+```
+
+We can then populate the Employee bean with the data provided in the following `applicationContext.xml` file:
+This makes our code loosely coupled and easier for testing. 
+```xml
+<bean id="employeeAmy" class="com.company.Employee">  
+    <property name="id">  
+        <value>20</value>  
+    </property>  
+    <property name="name">  
+        <value>Amy</value>  
+    </property>  
+</bean>  
+```
+   
+To better understand how this dependency injection works, we can have a `Test.java` looks like the following:
```

Comment:
Also consider this flow:
First, give the employee code.
Then, give the xml file that explains how to populate employee objects.
Then, show how the xml file is used to instantiate the container, which will result in objects being populated.
After that, show the test code, which proves the objects have been populated.